tools/c7n_mailer/c7n_mailer/gcp/pubsub_subscriptions.py

```diff
@@ -0,0 +1,184 @@
+            subscription)
+
+    @staticmethod
+    def ensure_subscription(session, data, subscription=None):
+        """Verify the pub/sub topic exists.
+        If it does not, create it returns the topic qualified name.
+        """
+
+        client = session.client('pubsub', 'v1', 'projects.subscriptions')
+
+        topic = PubSubSubscriptions.ensure_topic(session, data)
+        subscription = PubSubSubscriptions.get_subscription_param(session, subscription)
+
+        try:
+            client.execute_command('get', {'subscription': subscription})
+        except HttpError as e:
+            if e.resp.status != 404:
+                raise
+        else:
+            return subscription
+
+        # bug in discovery doc.. apis say body must be empty but its required in the
+        # discovery api for create.
+        client.execute_command('create', {'name': subscription, 'body': {}})
+        return topic
+
+        pass
+
+    """
+    Class methods for handling Pub/Sub Topics
+    """
+    @staticmethod
+    def get_topic_param(session, data, topic=None, project=None):
+        """Returns Rest API URI formatted with topic and project in it
+        """
+        return 'projects/{}/topics/{}'.format(
+            project or session.get_default_project(),
+            topic or data['transport']['topic'])
+
+    @staticmethod
+    def ensure_topic(session, data):
+        """Verify the pub/sub topic exists.
+        If it does not, create it
+
+        Returns the topic qualified name.
+        """
+
+        client = session.client('pubsub', 'v1', 'projects.topics')
+
+        topic = PubSubSubscriptions.get_topic_param(session, data)
+        try:
+            client.execute_command('get', {'topic': topic})
+        except HttpError as e:
+            if e.resp.status != 404:
+                raise
+        else:
+            return topic
+
+        # bug in discovery doc.. apis say body must be empty but its required in the
+        # discovery api for create.
+        client.execute_command('create', {'name': topic, 'body': {}})
+        return topic
+
+    @staticmethod
+    def ensure_iam(session, data, publisher=None):
+        """ Ensures that the correct Iam Permissions are setup for a topic
+        """
+        topic = PubSubSubscriptions.get_topic_param(session, data)
+        client = session.client('pubsub', 'v1', 'projects.topics')
+        policy = client.execute_command('getIamPolicy', {'resource': topic})
+        policy.pop('etag')
+        found = False
+        for binding in policy.get('bindings', {}):
+            if binding['role'] != 'roles/pubsub.publisher':
+                continue
+            if publisher in binding['members']:
+                return
+            found = binding
+
+        if not found:
+            policy.setdefault(
+                'bindings', {'members': [publisher], 'role': 'roles/pubsub.publisher'})
+        else:
+            found['members'].append(publisher)
```

Comment:
how does this work for capturing multiple bindings from the policy if it contains mulitple? Because it's only going to happen on the last item of the loop since this append is in a check outside of the loop. I think some of the execution order of the for-loop might should be re-evaluated, or I might just be seeing it weird. But it just seems like there could be some weird scenarios happening because of the order of when conditionals are checked and variables are assigned. this would be a little easier to explain in person.